### PR TITLE
nvmem: raspberrypi: Add nvmem driver for reading customer OTP

### PR DESCRIPTION
**Pull request by popcornmix**
This supports reading and writing OTP using the firmware mailbox interface.

It needs supporting firmware to run.

### CODE REVIEW

File: arch/arm/boot/dts/broadcom/bcm2712-rpi.dtsi [hunk 1]
```diff
@@ -45,6 +45,34 @@
...
+	};
+
+	nvmem_customer: nvmem_customer {
+		compatible = "raspberrypi,rpi-otp";
+		firmware = <&firmware>;
+		reg = <1 8 1>;
+		status = "okay";
+	};
+
+	nvmem_mac: nvmem_mac {
```
Comment (timg236) [inline arch/arm/boot/dts/broadcom/bcm2712-rpi.dtsi:62]: Is it possible for a user to make the MAC addresses visible but read-only via device-tree or module parameters?

**Comment by popcornmix** [comment]
@timg236 updated with:
```
dtparam=nvmem_customer_writeable
```
Allow write access to customer otp through nvmem driver
(I suspect we will want to change names, and probably where these entries are places in DT files).

File: drivers/nvmem/Kconfig [hunk 2]
```diff
@@ -39,6 +39,19 @@ config NVMEM_APPLE_EFUSES
 	  This driver can also be built as a module. If so, the module will
 	  be called nvmem-apple-efuses.
 
+config NVMEM_RASPBERRYPI_OTP
+	tristate "Raspberry Pi OTP support"
+	depends on ARCH_BCM || COMPILE_TEST
```
Comment (pelwell) [inline drivers/nvmem/Kconfig]: I think this should depend on or select RASPBERRYPI_FIRMWARE.

**Comment by popcornmix** [comment]
@pelwell @timg236 I think this is ready for review.

**Comment by timg236** [comment]
> @pelwell @timg236 I think this is ready for review.

Thanks I’ll try this out tomorrow

File: arch/arm/boot/dts/overlays/README [hunk 3]
```diff
@@ -300,6 +300,12 @@ Params:
 
         nvme                    Alias for "pciex1" (2712 only)
 
+        nvmem_customer_writeable Allow write access to customer otp through nvmem driver
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README]: The parameter decription must start on column 33, and lines must not exceed 80 characters.

File: drivers/nvmem/raspberrypi-otp.c [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	int words = bytes / sizeof(u32);
+	int index = offset / sizeof(u32);
+	u32 data[3 + MAX_ROWS] = {priv->block, index, words};
+	int err = 0;
+
+	if (bytes > MAX_ROWS * sizeof(u32))
+		return -EINVAL;
+
+	memcpy(data + 3, val, bytes);
+	err = rpi_firmware_property(priv->fw, RPI_FIRMWARE_SET_USER_OTP,
```
Comment (pelwell) [inline drivers/nvmem/raspberrypi-otp.c]: Is there any reason not to drop err and just "return rpi_firmware_property(..."?

File: drivers/nvmem/Kconfig [hunk 5]
```diff
@@ -39,6 +39,19 @@ config NVMEM_APPLE_EFUSES
 	  This driver can also be built as a module. If so, the module will
 	  be called nvmem-apple-efuses.
 
+config NVMEM_RASPBERRYPI_OTP
+	tristate "Raspberry Pi OTP support"
+	depends on (ARCH_BCM && RASPBERRYPI_FIRMWARE) || COMPILE_TEST
+	default ARCH_BCM && RASPBERRYPI_FIRMWARE
+	help
+	  Say y here to enable support for reading OTP on Raspberry Pi boards.
+	  such as the M1. These are e.g. used to store non-volatile board
```
Comment (pelwell) [inline drivers/nvmem/Kconfig]: "such as the M1." is unwanted text copied from the entry above.

**Review by pelwell — COMMENTED** [review]
Apart from the inline comments, this needs to be split into two commits.

File: arch/arm/boot/dts/overlays/README (hunk 3 above)
Comment (popcornmix) [reply arch/arm/boot/dts/overlays/README]: The parameter name is too long for that. Any preference for a change?
`nvmem_customer_w` or `customer_writeable`?

**Comment by pelwell** [comment]
I think you can cover all downstream dtbs by putting the changes in `bcm2708-rpi.dtsi`, `bcm2711-rpi-ds.dtsi` and `bcm2712-rpi.dtsi`. If  we want to change upstream as well, move them to `bcm2835-common.dtsi` and `bcm2711-rpi.dtsi`.

**Comment by pelwell** [comment]
I also notice that the 2712 nvmem entries are in `/soc`, whereas the others are in `/`.

**Comment by pelwell** [comment]
See https://github.com/pelwell/linux/tree/nvmem - the output is identical. Also consider moving the 2712 nvmem nodes to /

**Comment by popcornmix** [comment]
Updated.

File: arch/arm/boot/dts/overlays/README [hunk 6]
```diff
@@ -300,6 +300,12 @@ Params:
 
         nvme                    Alias for "pciex1" (2712 only)
 
+        nvmem_cust_rw           Allow read/write access to cust otp through nvmem driver
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README]: These are over 80 characters long

**Review by pelwell — COMMENTED** [review]
> Also consider moving the 2712 nvmem nodes to /

Actually, moving the non-2712 nvmems to /soc removes a dtc warning about reg property length.

**Comment by pelwell** [comment]
```diff
diff --git a/arch/arm/boot/dts/broadcom/bcm2708-rpi.dtsi b/arch/arm/boot/dts/broadcom/bcm2708-rpi.dtsi
index 4302f55f57b4d..f4949a07272b6 100644
--- a/arch/arm/boot/dts/broadcom/bcm2708-rpi.dtsi
+++ b/arch/arm/boot/dts/broadcom/bcm2708-rpi.dtsi
@@ -15,6 +15,17 @@ aliases {
                i2c2 = &i2c2;
        };
 
+       __overrides__ {
+               hdmi = <&hdmi>,"status";
+               i2c2_iknowwhatimdoing = <&i2c2>,"status";
+               i2c2_baudrate = <&i2c2>,"clock-frequency:0";
+               nvmem_cust_rw = <&nvmem_cust>,"rw?";
+               sd = <&sdhost>,"status";
+               sd_poll_once = <&sdhost>,"non-removable?";
+       };
+};
+
+&soc {
        nvmem_otp: nvmem_otp {
                compatible = "raspberrypi,rpi-otp";
                firmware = <&firmware>;
@@ -28,15 +39,6 @@ nvmem_cust: nvmem_cust {
                reg = <1 8>;
                status = "okay";
        };
-
-       __overrides__ {
-               hdmi = <&hdmi>,"status";
-               i2c2_iknowwhatimdoing = <&i2c2>,"status";
-               i2c2_baudrate = <&i2c2>,"clock-frequency:0";
-               nvmem_cust_rw = <&nvmem_cust>,"rw?";
-               sd = <&sdhost>,"status";
-               sd_poll_once = <&sdhost>,"non-removable?";
-       };
 };
 
 &sdhost {
diff --git a/arch/arm/boot/dts/broadcom/bcm2711-rpi-ds.dtsi b/arch/arm/boot/dts/broadcom/bcm2711-rpi-ds.dtsi
index daa3b5dfa0bf3..b298cd8bb71ce 100644
--- a/arch/arm/boot/dts/broadcom/bcm2711-rpi-ds.dtsi
+++ b/arch/arm/boot/dts/broadcom/bcm2711-rpi-ds.dtsi
@@ -75,6 +75,21 @@ zone_dma {
                #size-cells = <1>;
                dma-ranges = <0x0  0x0 0x0  0x40000000>;
        };
+};
+
+&vc4 {
+       raspberrypi,firmware = <&firmware>;
+};
+
+&cma {
+       /* Limit cma to the lower 768MB to allow room for HIGHMEM on 32-bit */
+       alloc-ranges = <0x0 0x00000000 0x30000000>;
+};
+
+&soc {
+       /* Add the physical <-> DMA mapping for the I/O space */
+       dma-ranges = <0xc0000000  0x0 0x00000000  0x40000000>,
+                    <0x7c000000  0x0 0xfc000000  0x03800000>;
 
        nvmem_otp: nvmem_otp {
                compatible = "raspberrypi,rpi-otp";
@@ -98,21 +113,6 @@ nvmem_priv: nvmem_priv {
        };
 };
 
-&vc4 {
-       raspberrypi,firmware = <&firmware>;
-};
-
-&cma {
-       /* Limit cma to the lower 768MB to allow room for HIGHMEM on 32-bit */
-       alloc-ranges = <0x0 0x00000000 0x30000000>;
-};
-
-&soc {
-       /* Add the physical <-> DMA mapping for the I/O space */
-       dma-ranges = <0xc0000000  0x0 0x00000000  0x40000000>,
-                    <0x7c000000  0x0 0xfc000000  0x03800000>;
-};
-
 &scb {
        #size-cells = <2>;
 
```

**Comment by popcornmix** [comment]
Updated

**Pull request merged**